Add some IOCTLs for evdev

Cautreoxit opened the pull request:
This PR is to add some IOCTLs for `evdev`, which are necessary for `xorg server`.

Actually the first commit is from evdev (PR https://github.com/asterinas/asterinas/pull/2561/), and the second commit is for IOCTLs. Since evdev has not been merged yet, I mark this PR as _draft_. After evdev has final version, I would modify this PR and mark it _open_.

Cautreoxit commented:
The reason for the CI test failure is that the kernel API used by the evdev module has changed in the past few days. Simply updating the evdev module will resolve the issue, which has already been addressed in the evdev PR. These errors are unrelated to ioctl itself.

Cautreoxit commented:
Since evdev PR (https://github.com/asterinas/asterinas/pull/2561/) has been merged, I modified this PR and made it ready for review.

tatetian commented on kernel/src/device/evdev/file.rs:
```diff
@@ -185,6 +214,192 @@ impl EvdevFile {
+        match Self::parse_evdev(raw) {
+            Some(EvdevIoctl::GetName { len }) => {
+                let evdev = self.upgrade_evdev_device()?;
+                self.write_string_to_userspace(evdev.device.name(), len as usize, arg)?;
```
The `write_string_to_userspace` and `write_bitmap_to_userspace` methods should NOT have been made as methods of `EvdevFile`. A method accesses `self`. If it does not need `self`, then it should not be a method. Turn them into two functions.

tatetian commented on kernel/src/device/evdev/file.rs:
```diff
@@ -185,6 +214,192 @@ impl EvdevFile {
+            return Ok(());
+        }
+
+        let mut buffer = vec![0u8; len];
```
Optimize this function to avoid heap allocation and an extra data copy.

tatetian commented on kernel/src/device/evdev/file.rs:
```diff
@@ -185,6 +214,192 @@ impl EvdevFile {
+                    t if t == EventTypes::REL.as_index() => {
+                        Some(capability.supported_relative_axes_bitmap())
+                    }
+                    _ => None,
```
If `t` is not a recognized type, shouldn't we return an error?

tatetian commented:
@Cautreoxit The revised ioctl is still not quite satisfying. Check out #2649 for a complete design. It addresses some limitations of the ioctl infra in this PR.

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -211,6 +326,91 @@ impl EvdevFile {
+        } else if let Ok(ioctl) = SetClockId::try_from(request) {
+            let clock_id_raw: i32 = read_val_from_userspace(ioctl.user_ptr())?;
+            let clock_id = ClockId::try_from(clock_id_raw)
+                .map_err(|_| Error::with_message(Errno::EINVAL, "invalid clock id"))?;
```
`ClockId::try_from(clock_id_raw)?` should be okay.

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -211,6 +326,91 @@ impl EvdevFile {
+                    return_errno_with_message!(Errno::EINVAL, "unsupported clock id");
+                }
+            };
+            self.inner.clock_id.store(evdev_clock, Ordering::Relaxed);
```
A logic to flush the queue is missed here.

Honestly, I don't really know what's the best way to implement this... Something like this should work but it is quite inelegant:
```patch
diff --git a/kernel/src/device/evdev/file.rs b/kernel/src/device/evdev/file.rs
index 534680f2c..6511909d7 100644
--- a/kernel/src/device/evdev/file.rs
+++ b/kernel/src/device/evdev/file.rs
@@ -131,6 +131,28 @@ impl EvdevFileInner {
         self.pollee.invalidate();
     }
 
+    pub(self) fn clear_with_consumer_producer_locked(
+        &self,
+        consumer: &mut RbConsumer<EvdevEvent>,
+        producer: &mut RbProducer<EvdevEvent>,
+    ) {
+        // Note that the following operations are race-free because both the consumer and the
+        // producer locks have been held.
+
+        if consumer.len() == 0 {
+            return;
+        }
+
+        consumer.clear();
+
+        let dropped_event =
+            EvdevEvent::from_event_and_time(&InputEvent::from_sync_event(SynEvent::Dropped), time);
+        producer.push(dropped_event);
+
+        self.packet_count.store(1, Ordering::Relaxed);
+        self.pollee.notify(IoEvents::IN);
+    }
+
     /// Checks if buffer has complete event packets.
     pub(self) fn has_complete_packets(&self) -> bool {
         self.packet_count.load(Ordering::Relaxed) > 0
@@ -214,6 +236,30 @@ impl EvdevFile {
             IoEvents::empty()
         }
     }
+
+    fn upgrade_device(&self) -> Result<Arc<EvdevDevice>> {
+        self.evdev.upgrade().ok_or_else(|| {
+            Error::with_message(Errno::ENODEV, "the evdev device has been disconnected")
+        })
+    }
+
+    fn set_clock(&self, clock: EvdevClock) -> Result<()> {
+        // Lock the mutex in advance to avoid race conditions.
+        let mut consumer = self.inner.consumer.lock();
+
+        if self.inner.clock_id.load(Ordering::Relaxed) == clock {
+            return Ok(());
+        }
+
+        let device = self.upgrade_device()?;
+        device.with_producer_locked(&self.inner, |producer| {
+            self.inner.clock_id.store(clock, Ordering::Relaxed);
+            self.inner
+                .clear_with_consumer_producer_locked(&mut consumer, producer);
+        });
+
+        Ok(())
+    }
 }
 
 /// Checks if the event is a `SYN_REPORT` event.
diff --git a/kernel/src/device/evdev/mod.rs b/kernel/src/device/evdev/mod.rs
index cd2319cfb..6ef8e53a5 100644
--- a/kernel/src/device/evdev/mod.rs
+++ b/kernel/src/device/evdev/mod.rs
@@ -108,6 +108,18 @@ impl EvdevDevice {
         opened_files.swap_remove(pos);
     }
 
+    pub(self) fn with_producer_locked<F>(&self, file: &Arc<EvdevFileInner>, f: F)
+    where
+        F: FnOnce(&mut RbProducer<EvdevEvent>),
+    {
+        let mut opened_files = self.opened_files.disable_irq().lock();
+        let pos = opened_files
+            .iter()
+            .position(|(f, _)| Arc::ptr_eq(f, file))
+            .unwrap();
+        f(&mut opened_files[pos].1)
+    }
+
     /// Distributes events to all opened evdev files.
     fn pass_events(&self, events: &[InputEvent]) {
         // No need to disable IRQs because this method can only be called in the interrupt context.
```

I am curious if there is a better way...

Cautreoxit replied:
Yes, you're right — the current implementation indeed misses updating the event timestamp. The fix is admittedly a little inelegant, but I think it is the best approach for this situation. I’m using this solution for now.

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -211,6 +326,91 @@ impl EvdevFile {
+                t if t == EventTypes::KEY.as_index() => capability.supported_keys_bitmap(),
+                t if t == EventTypes::REL.as_index() => capability.supported_relative_axes_bitmap(),
+                _ => {
+                    return_errno_with_message!(Errno::EINVAL, "unsupported event type");
```
```rust
                    return_errno_with_message!(Errno::EINVAL, "the event type is not supported yet");
```

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -211,6 +326,91 @@ impl EvdevFile {
+        } else if let Ok(ioctl) = GetKeyState::try_from(request) {
+            let zero = vec![0u8; ioctl.buffer_len()];
+            write_bitmap_to_userspace(ioctl.user_ptr(), &zero[..], ioctl.buffer_len())?;
+            Ok(())
```
This is really just incorrect. This needs to return whether the key is currently pressed (and deal with the queue stuffs: https://elixir.bootlin.com/linux/v6.15/source/drivers/input/evdev.c#L872-L876).

At least, this needs a detailed TODO comment to explain the Linux behavior, and say something about the current implementation incorrectness.

Cautreoxit replied:
Yes. I haven’t actually implemented `GetKeyState` yet because it would require tracking the state of all keys, which we don’t currently support. Our target application, `Xorg` (via `libevdev`), does not use the information returned by `GetKeyState` (or `GetLedState` and `GetSwState`). So I left a dummy implementation in place and I have added a TODO comment here.

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -211,6 +326,91 @@ impl EvdevFile {
+            write_bitmap_to_userspace(ioctl.user_ptr(), &zero[..], ioctl.buffer_len())?;
+            Ok(())
+        } else if let Ok(ioctl) = GetLedState::try_from(request) {
+            let zero = vec![0u8; ioctl.buffer_len()];
```
Use `fill_zeros` here should be fine?

Cautreoxit commented:
Thanks for the thoughtful review! I’ve rebased this PR and addressed the comments above. I also responded to the two issues individually. If anything still looks off, please let me know. Thanks!

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -27,14 +28,35 @@ use crate::{
+    pub(super) type GetLedState   = ioc!(EVIOCGLED,     b'E', 0x19, OutData<[u8]>);
+    pub(super) type GetSwState    = ioc!(EVIOCGSW,      b'E', 0x1B, OutData<[u8]>);
+
+    pub(super) type GetBit = IoctlEnum<b'E', 0x20, 0x1F, OutData<[u8]>>;
```
Please mention the Linux name otherwise it's hard to look it up from the reference:
```rust
/// The `EVIOCGBIT` ioctl enum.
/// type GetEventBits = IoctlEnum<b'E', 0x20, 0x1F, OutData<[u8]>>;
```

Besides, the example in `ioctl.rs` uses `GetEventBits`, I think it's clearer so let's use it here as well?

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -131,6 +169,29 @@ impl EvdevFileInner {
+        let time = self.read_clock();
+        let dropped_event =
+            EvdevEvent::from_event_and_time(&InputEvent::from_sync_event(SynEvent::Dropped), time);
+        producer.push(dropped_event);
```
```suggestion
        producer.push(dropped_event).unwrap();
```

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -292,17 +427,88 @@ impl FileIo for EvdevFile {
+                })?;
+            }
+            cmd @ GetSwState => {
+                // No SW events are supported. So we can report zeros as switch states here.
```
```suggestion
                // No switch events are supported. So we can report zeros as switch states here.
```

lrh2000 commented on kernel/src/device/evdev/file.rs:
```diff
@@ -292,17 +427,88 @@ impl FileIo for EvdevFile {
+                // user space's state tracking. See the Linux implementation at:
+                // <https://elixir.bootlin.com/linux/v6.15/source/drivers/input/evdev.c#L872-L876>.
+
+                // Currently, no key state tracking is supported. So we report zeros as key states here.
```
```suggestion
                // <https://elixir.bootlin.com/linux/v6.15/source/drivers/input/evdev.c#L872-L876>.
                //
                // Currently, no key state tracking is supported. So we report zeros as key states here.
```

lrh2000 reviewed (approved):
Thanks for the revision. I took another look and have a few more comments, but they're minor style issues. I'm approving this, and then @tatetian can perform the final checks.

Cautreoxit commented:
Thanks for your comments and approval! I have addressed these issues now.

Regarding the question in this comment https://github.com/asterinas/asterinas/pull/2582#discussion_r2596137771, yes, `libevdev` calls the `GetKeyState` ioctl during initialization, although our application doesn't use the information returned. However, it does check the result code, and if there's an error, the initialization would fail. Therefore, we cannot allow these ioctls to fail. Here's how `evdevlib` uses `GetKeyState`:
https://github.com/whot/libevdev/blob/c3953e1bb8f813f3e248047d23fb0775b147da9f/libevdev/libevdev.c#L428
